distributor/cmd/internal/distributor/distributor.go

```diff
@@ -0,0 +1,244 @@
+				break
+			}
+			cpsAtSize = make([][]byte, 0)
+			witsAtSize = make([]note.Verifier, 0)
+			currentSize = size
+		}
+		witsAtSize = append(witsAtSize, d.ws[witID])
+		cpsAtSize = append(cpsAtSize, cp)
+	}
+	if err := rows.Err(); err != nil {
+		return nil, fmt.Errorf("failed to scan rows: %v", err)
+	}
+	if len(cpsAtSize) >= int(n) {
+		cp, err := checkpoints.Combine(cpsAtSize, l.Verifier, note.VerifierList(witsAtSize...))
+		if err != nil {
+			return nil, fmt.Errorf("failed to combine sigs: %v", err)
+		}
+		return cp, nil
+	}
+	return nil, fmt.Errorf("no checkpoint with %d signatures found", n)
+}
+
+// GetCheckpointWitness gets the largest checkpoint for the log that was witnessed by the given witness.
+func (d *Distributor) GetCheckpointWitness(ctx context.Context, logID, witID string) ([]byte, error) {
+	tx, err := d.db.BeginTx(ctx, &sql.TxOptions{ReadOnly: true})
+	if err != nil {
+		return nil, fmt.Errorf("failed to begin transaction: %v", err)
+	}
+	return getLatestCheckpoint(ctx, tx, logID, witID)
+}
+
+// Distribute adds a new witnessed checkpoint to be distributed. This checkpoint must be signed
+// by both the log and the witness specified, and be larger than any previous checkpoint distributed
+// for this pair.
+func (d *Distributor) Distribute(ctx context.Context, logID, witID string, nextRaw []byte) error {
+	l, ok := d.ls[logID]
+	if !ok {
+		return fmt.Errorf("unknown log ID %q", logID)
+	}
+	wv, ok := d.ws[witID]
+	if !ok {
+		return fmt.Errorf("unknown witness ID %q", witID)
+	}
+	newCP, _, n, err := log.ParseCheckpoint(nextRaw, l.Origin, l.Verifier, wv)
+	if err != nil {
+		return fmt.Errorf("failed to parse checkpoint: %v", err)
+	}
+	if len(n.Sigs) != 2 {
+		return fmt.Errorf("failed to verify log and witness signatures; only verified: %v", n.Sigs)
+	}
+
+	// This is a valid checkpoint for this log for this witness
+	// Now find the previous checkpoint if one exists.
+
+	tx, err := d.db.BeginTx(ctx, &sql.TxOptions{ReadOnly: false})
+	if err != nil {
+		return fmt.Errorf("failed to begin transaction: %v", err)
+	}
+	saveCheckpointFn := func() error {
+		_, err := tx.ExecContext(ctx, `INSERT OR REPLACE INTO chkpts (logID, witID, treeSize, chkpt) VALUES (?, ?, ?, ?)`, logID, witID, newCP.Size, nextRaw)
+		if err != nil {
+			return fmt.Errorf("Exec(): %v", err)
+		}
+		if err := tx.Commit(); err != nil {
+			return err
+		}
+		return nil
+	}
+	oldBs, err := getLatestCheckpoint(ctx, tx, logID, witID)
+	if err != nil {
+		if status.Code(err) == codes.NotFound {
+			// If this is the first checkpoint for this witness then just save and exit
+			return saveCheckpointFn()
+		}
+		return err
+	}
+
+	// We have the previous checkpoint, now check that the new one is fresher
+
+	oldCP, _, _, err := log.ParseCheckpoint(oldBs, l.Origin, l.Verifier, wv)
+	if err != nil {
+		// This really shouldn't ever happen unless the DB is corrupted or the config
+		// for the log or verifier has changed.
+		return err
```

Comment:
Might be worth "decorating" this error to be clear it's the one from the DB (mirroring the decorated error above when parsing the passed-in CP)?

Reply:
ta, done